## Here the results for all the exoeriments awill be evaluated final models will be trained on train+val data and then fitted and checked with the test data, and finally will get the plots.

In [1]:
import torch
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly as pt
import sys
import os

sys.path.append(os.path.abspath(".."))
os.makedirs("../results/final_models",exist_ok=True)

In [2]:
baselines = pd.read_csv("results/baseline_results.csv")
baselines

,model_type,val_smape,test_smape,name,seq_len,pred_len,n_heads,dropout,d_model,enc_layers,model_smape,naive_smape,beats_naive
0,ARIMA,0.624492,0.705411,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,transformer,NaN,NaN,transformer_s192_p199_h4_k23_d0.1,192.0,199.0,4.0,0.1,64.0,2.0,0.187262,1.677571,True
2,transformer,NaN,NaN,transformer_s192_p199_h5_k23_d0.1,192.0,199.0,5.0,0.1,80.0,2.0,0.210449,1.677571,True
3,transformer,NaN,NaN,transformer_s256_p199_h5_k23_d0.1,256.0,199.0,5.0,0.1,80.0,2.0,0.208822,1.708083,True
4,transformer,NaN,NaN,transformer_s256_p199_h5_k23_d0.2,256.0,199.0,5.0,0.2,80.0,2.0,0.211516,1.708083,True
5,transformer,NaN,NaN,transformer_s256_p199_h5_k23_d0.2,256.0,199.0,5.0,0.2,80.0,2.0,0.402456,1.615023,True


In [3]:
experiments=pd.read_csv("results/experiment_results.csv")
experiments

,model_type,seq_len,pred_len,n_heads,dropout,d_model,enc_layers,name,model_smape,naive_smape,beats_naive
0,informer,48,24,3,0.1,60,2,informer_s48_p24_h3_k23_d0.1,0.086753,1.852614,True
1,informer,48,96,3,0.1,60,2,informer_s48_p96_h3_k23_d0.1,0.149948,1.726145,True
2,informer,48,199,3,0.1,60,2,informer_s48_p199_h3_k23_d0.1,0.252092,1.623904,True
3,informer,48,366,3,0.1,60,2,informer_s48_p366_h3_k23_d0.1,0.309085,1.446866,True
4,informer,96,24,3,0.1,60,2,informer_s96_p24_h3_k23_d0.1,0.085383,1.857614,True
...,...,...,...,...,...,...,...,...,...,...,...
63,autoformer,256,366,5,0.1,80,2,autoformer_s256_p366_h5_k23_d0.1,0.239401,1.614763,True
64,autoformer,96,199,5,0.2,80,2,autoformer_s96_p199_h5_k23_d0.2,0.180204,1.657231,True
65,autoformer,96,199,5,0.2,80,2,autoformer_s96_p199_h5_k11_d0.2,0.183465,1.657231,True
66,autoformer,96,199,5,0.2,80,2,autoformer_s96_p199_h5_k23_d0.2,0.292093,1.576994,True


### The columns for the tables dont exactly match so some cleaning is required 

In [4]:
baselines.loc[0,"name"]="ARIMA"
baselines.loc[0,"seq_len"]=None
baselines.loc[0,"pred_len"]=None
baselines.loc[0,"n_heads"]=None
baselines.loc[0,"dropout"]=None
baselines.loc[0,"d_model"]=None


In [5]:
for i in range (1,6,1):
    baselines.loc[i,"val_smape"]=baselines.loc[i,"model_smape"] 
    
baselines

,model_type,val_smape,test_smape,name,seq_len,pred_len,n_heads,dropout,d_model,enc_layers,model_smape,naive_smape,beats_naive
0,ARIMA,0.624492,0.705411,ARIMA,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,transformer,0.187262,NaN,transformer_s192_p199_h4_k23_d0.1,192.0,199.0,4.0,0.1,64.0,2.0,0.187262,1.677571,True
2,transformer,0.210449,NaN,transformer_s192_p199_h5_k23_d0.1,192.0,199.0,5.0,0.1,80.0,2.0,0.210449,1.677571,True
3,transformer,0.208822,NaN,transformer_s256_p199_h5_k23_d0.1,256.0,199.0,5.0,0.1,80.0,2.0,0.208822,1.708083,True
4,transformer,0.211516,NaN,transformer_s256_p199_h5_k23_d0.2,256.0,199.0,5.0,0.2,80.0,2.0,0.211516,1.708083,True
5,transformer,0.402456,NaN,transformer_s256_p199_h5_k23_d0.2,256.0,199.0,5.0,0.2,80.0,2.0,0.402456,1.615023,True


In [6]:
baselines=baselines.drop(columns=["model_smape","naive_smape", "beats_naive"])
baselines

,model_type,val_smape,test_smape,name,seq_len,pred_len,n_heads,dropout,d_model,enc_layers
0,ARIMA,0.624492,0.705411,ARIMA,NaN,NaN,NaN,NaN,NaN,NaN
1,transformer,0.187262,NaN,transformer_s192_p199_h4_k23_d0.1,192.0,199.0,4.0,0.1,64.0,2.0
2,transformer,0.210449,NaN,transformer_s192_p199_h5_k23_d0.1,192.0,199.0,5.0,0.1,80.0,2.0
3,transformer,0.208822,NaN,transformer_s256_p199_h5_k23_d0.1,256.0,199.0,5.0,0.1,80.0,2.0
4,transformer,0.211516,NaN,transformer_s256_p199_h5_k23_d0.2,256.0,199.0,5.0,0.2,80.0,2.0
5,transformer,0.402456,NaN,transformer_s256_p199_h5_k23_d0.2,256.0,199.0,5.0,0.2,80.0,2.0


In [7]:
for i in range (0,68,1):
    experiments.loc[i,"val_smape"]=experiments.loc[i,"model_smape"]

experiments


,model_type,seq_len,pred_len,n_heads,dropout,d_model,enc_layers,name,model_smape,naive_smape,beats_naive,val_smape
0,informer,48,24,3,0.1,60,2,informer_s48_p24_h3_k23_d0.1,0.086753,1.852614,True,0.086753
1,informer,48,96,3,0.1,60,2,informer_s48_p96_h3_k23_d0.1,0.149948,1.726145,True,0.149948
2,informer,48,199,3,0.1,60,2,informer_s48_p199_h3_k23_d0.1,0.252092,1.623904,True,0.252092
3,informer,48,366,3,0.1,60,2,informer_s48_p366_h3_k23_d0.1,0.309085,1.446866,True,0.309085
4,informer,96,24,3,0.1,60,2,informer_s96_p24_h3_k23_d0.1,0.085383,1.857614,True,0.085383
...,...,...,...,...,...,...,...,...,...,...,...,...
63,autoformer,256,366,5,0.1,80,2,autoformer_s256_p366_h5_k23_d0.1,0.239401,1.614763,True,0.239401
64,autoformer,96,199,5,0.2,80,2,autoformer_s96_p199_h5_k23_d0.2,0.180204,1.657231,True,0.180204
65,autoformer,96,199,5,0.2,80,2,autoformer_s96_p199_h5_k11_d0.2,0.183465,1.657231,True,0.183465
66,autoformer,96,199,5,0.2,80,2,autoformer_s96_p199_h5_k23_d0.2,0.292093,1.576994,True,0.292093


In [8]:
experiments =experiments.drop(columns=["model_smape","naive_smape","beats_naive"])
experiments

,model_type,seq_len,pred_len,n_heads,dropout,d_model,enc_layers,name,val_smape
0,informer,48,24,3,0.1,60,2,informer_s48_p24_h3_k23_d0.1,0.086753
1,informer,48,96,3,0.1,60,2,informer_s48_p96_h3_k23_d0.1,0.149948
2,informer,48,199,3,0.1,60,2,informer_s48_p199_h3_k23_d0.1,0.252092
3,informer,48,366,3,0.1,60,2,informer_s48_p366_h3_k23_d0.1,0.309085
4,informer,96,24,3,0.1,60,2,informer_s96_p24_h3_k23_d0.1,0.085383
...,...,...,...,...,...,...,...,...,...
63,autoformer,256,366,5,0.1,80,2,autoformer_s256_p366_h5_k23_d0.1,0.239401
64,autoformer,96,199,5,0.2,80,2,autoformer_s96_p199_h5_k23_d0.2,0.180204
65,autoformer,96,199,5,0.2,80,2,autoformer_s96_p199_h5_k11_d0.2,0.183465
66,autoformer,96,199,5,0.2,80,2,autoformer_s96_p199_h5_k23_d0.2,0.292093


## Now let's start with the final model selection 

In [9]:
transformers=baselines[baselines["model_type"]=="transformer"]
informers=experiments[experiments["model_type"]=="informer"]
autoformers=experiments[experiments["model_type"]=="autoformer"]

### Here for choosing the best model i have kept the horizon to be constant =199 which is a long horizon but not too long as a year's gap (366)

In [10]:
autoformer_199p=autoformers[autoformers["pred_len"]==199]
best_autoformer=autoformer_199p.loc[autoformer_199p["val_smape"].idxmin()]
best_autoformer

model_type                         autoformer
seq_len                                    96
pred_len                                  199
n_heads                                     5
dropout                                   0.2
d_model                                    80
enc_layers                                  2
name          autoformer_s96_p199_h5_k23_d0.2
val_smape                            0.180204
Name: 64, dtype: object

In [11]:
informer_199p=informers[informers["pred_len"]==199]
best_informer=informer_199p.loc[informer_199p["val_smape"].idxmin()]
best_informer

model_type                         informer
seq_len                                  96
pred_len                                199
n_heads                                   3
dropout                                 0.1
d_model                                  60
enc_layers                                2
name          informer_s96_p199_h3_k23_d0.1
val_smape                          0.205611
Name: 6, dtype: object

In [12]:
transformer_199p=transformers[transformers["pred_len"]==199]
best_transformer=transformer_199p.loc[transformer_199p["val_smape"].idxmin()]
best_transformer

model_type                          transformer
val_smape                              0.187262
test_smape                                  NaN
name          transformer_s192_p199_h4_k23_d0.1
seq_len                                   192.0
pred_len                                  199.0
n_heads                                     4.0
dropout                                     0.1
d_model                                    64.0
enc_layers                                  2.0
Name: 1, dtype: object

In [13]:
arima_model=baselines.loc[0]
arima_model

model_type       ARIMA
val_smape     0.624492
test_smape    0.705411
name             ARIMA
seq_len            NaN
pred_len           NaN
n_heads            NaN
dropout            NaN
d_model            NaN
enc_layers         NaN
Name: 0, dtype: object

In [14]:
final_table=pd.DataFrame([arima_model,best_transformer,best_informer,best_autoformer])
final_table.reset_index(drop=True, inplace=True)

In [15]:
final_table

,model_type,val_smape,test_smape,name,seq_len,pred_len,n_heads,dropout,d_model,enc_layers
0,ARIMA,0.624492,0.705411,ARIMA,NaN,NaN,NaN,NaN,NaN,NaN
1,transformer,0.187262,NaN,transformer_s192_p199_h4_k23_d0.1,192.0,199.0,4.0,0.1,64.0,2.0
2,informer,0.205611,NaN,informer_s96_p199_h3_k23_d0.1,96.0,199.0,3.0,0.1,60.0,2.0
3,autoformer,0.180204,NaN,autoformer_s96_p199_h5_k23_d0.2,96.0,199.0,5.0,0.2,80.0,2.0


In [16]:
import plotly.graph_objects as go

models = ["Arima","Transformer","Informer","Autoformer"]
smape = final_table["val_smape"]

fig = go.Figure()

fig.add_trace(
    go.Bar(
        x=models,
        y=smape,
        text=[f"{v:.3f}" for v in smape],
        textposition="outside",
        marker=dict(
            color=["#2E86AB", "#E74C3C","#9ACD32","#800080"]
        )
    )
)

fig.update_layout(
    title="Model vs Smape Comparision",
    xaxis_title="Modesls",
    yaxis_title="Validation sMAPE",
    template="plotly_white",
    font=dict(size=16),
    width=850,
    height=500
)

# fig.add_annotation(
#     x=1,
#     y=smape[1],
#     text="Loss of cross-currency information",
#     showarrow=True,
#     arrowhead=2
# )

fig.show()


In [17]:
import numpy as np
import torch
from src.data.window import create_windows
from src.data.dataloader import create_dataloaders
from src.models.build import build_model
from src.training.train import train_model
from src.evaluation.metrics import smape
from src.data.normalization import fit_scaler,transform_data
import pandas as pd

def train_best_model(best_row, train_data, val_data, test_data):

    device = torch.device("mps" if torch.backends.mps.is_available() else "cpu")

   
    # 1 Merge train + val
    train_val = np.vstack([train_data, val_data])

    # --- normalization ---
    scaler = fit_scaler(train_val)
    train_val = transform_data(train_val, scaler)
    test_data = transform_data(test_data, scaler)

    seq_len = int(best_row["seq_len"])
    pred_len = int(best_row["pred_len"])

    
    # 2 Create windows
    X_train, Y_train = create_windows(train_val, seq_len, pred_len)
    X_test, Y_test = create_windows(test_data, seq_len, pred_len)

    
    # 3 Residual targets
    Y_train_res = Y_train - X_train[:, -1:, :]
    Y_test_res = Y_test - X_test[:, -1:, :]

    
    # 4 Dataloaders
    train_loader, _ = create_dataloaders(
        X_train,
        Y_train_res,
        X_train,
        Y_train_res,
        batch_size=16
    )


    # 5 Build config
    config = {
        "model_type": best_row["model_type"],
        "seq_len": seq_len,
        "pred_len": pred_len,
        "n_heads": int(best_row["n_heads"]),
        "dropout": float(best_row["dropout"]),
        "d_model": int(best_row["d_model"]),
        "enc_layers": int(best_row["enc_layers"]),
        "input_dim": train_val.shape[1],
        "epochs": 20,
        "save_path": f"results/{best_row['model_type']}_final.pth",
        "lr":0.0001,
        "patience":5,
        "residual":True,
        "univariate":False,
        "kernel_size":23,
        "use_scaling":True
    }

    
    # 6 Build model
    model = build_model(config, device)

    # 7 Train
    train_model(
        model,
        train_loader,
        train_loader,
        config,
        device
    )
    # saving the model
    model_name=config["model_type"]
    save_path=f"../results/final_models/{model_name}_final.pth"
    torch.save(
    model.state_dict(),
    save_path
    )
    

    # 8 Test evaluation
    model.eval()
    test_smape = 0
    actual_list=[]
    pred_list=[]
    with torch.no_grad():

        for i in range(len(X_test)):

            x = torch.tensor(X_test[i:i+1], dtype=torch.float32).to(device)
            y_res = torch.tensor(Y_test_res[i:i+1], dtype=torch.float32).to(device)

            residual_pred = model(x)

            last_value = x[:, -1:, :]
            final_pred = last_value + residual_pred
            y_true = last_value + y_res

            test_smape += smape(y_true, final_pred).item()
            actual_list.append(y_true.cpu().numpy().flatten())
            pred_list.append(final_pred.cpu().numpy().flatten())
    test_smape /= len(X_test)
    actual=np.concatenate(actual_list)
    pred=np.concatenate(pred_list)

    
    forecast_df = pd.DataFrame({
        "time": np.arange(len(actual)),
        "actual": actual,
        config["model_type"]: pred
    })

    # simple uncertainty band
    forecast_df[f"{config['model_type']}_upper"] = pred + 0.02
    forecast_df[f"{config['model_type']}_lower"] = pred - 0.02

    save_path = f"results/{config['model_type']}_forecast.csv"

    forecast_df.to_csv(save_path, index=False)

    print(f"Forecast saved to {save_path}")
    


    return test_smape


In [18]:
from src.data.load_data import load_exchange_data

train_data,val_data,test_data=load_exchange_data(univariate=False)

/Users/akarn/Documents/Exchange_curr_pred/.venv/lib/python3.14/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [21]:
transformer_test=train_best_model(
    best_transformer,
    train_data,
    val_data,
    test_data
)

informer_test=train_best_model(
    best_informer,
    train_data,
    val_data,
    test_data
)
autoformer_test=train_best_model(
    best_autoformer,
    train_data,
    val_data,
    test_data
)

Epoch 1 | Train Loss 0.2481 | Val sMAPE 0.6541
Epoch 2 | Train Loss 0.1870 | Val sMAPE 0.7710
Epoch 3 | Train Loss 0.1733 | Val sMAPE 0.7353
Epoch 4 | Train Loss 0.1696 | Val sMAPE 0.6637
Epoch 5 | Train Loss 0.1777 | Val sMAPE 0.6997
Epoch 6 | Train Loss 0.1551 | Val sMAPE 0.6957
Early stopping triggered.
Forecast saved to results/transformer_forecast.csv


/Users/akarn/Documents/Exchange_curr_pred/src/models/informer.py:22: UserWarning: enable_nested_tensor is True, but self.use_nested_tensor is False because encoder_layer.self_attn.num_heads is odd
  self.encoder = nn.TransformerEncoder(


Epoch 1 | Train Loss 0.1847 | Val sMAPE 0.7686
Epoch 2 | Train Loss 0.1813 | Val sMAPE 0.6543
Epoch 3 | Train Loss 0.1779 | Val sMAPE 0.7065
Epoch 4 | Train Loss 0.1778 | Val sMAPE 0.6553
Epoch 5 | Train Loss 0.1555 | Val sMAPE 0.7956
Epoch 6 | Train Loss 0.1486 | Val sMAPE 0.8200
Epoch 7 | Train Loss 0.1373 | Val sMAPE 0.8810
Early stopping triggered.
Forecast saved to results/informer_forecast.csv


/Users/akarn/Documents/Exchange_curr_pred/src/models/autoformer.py:51: UserWarning: enable_nested_tensor is True, but self.use_nested_tensor is False because encoder_layer.self_attn.num_heads is odd
  self.encoder = nn.TransformerEncoder(


Epoch 1 | Train Loss 0.2488 | Val sMAPE 0.5999
Epoch 2 | Train Loss 0.1991 | Val sMAPE 0.6603
Epoch 3 | Train Loss 0.1923 | Val sMAPE 0.6777
Epoch 4 | Train Loss 0.1840 | Val sMAPE 0.6928
Epoch 5 | Train Loss 0.1776 | Val sMAPE 0.7018
Epoch 6 | Train Loss 0.1735 | Val sMAPE 0.7184
Early stopping triggered.
Forecast saved to results/autoformer_forecast.csv


### Feature normalization was introduced using StandardScaler fit on the training data and applied to validation/test sets. While normalization stabilized training, it did not significantly reduce the validation-to-test generalization gap. This suggests that the gap is primarily caused by temporal distribution shift in the exchange-rate dataset rather than feature scale differences.

In [23]:
final_table.loc[1,"test_smape"]=transformer_test
final_table.loc[2,"test_smape"]=informer_test
final_table.loc[3,"test_smape"]=autoformer_test
final_table

,model_type,val_smape,test_smape,name,seq_len,pred_len,n_heads,dropout,d_model,enc_layers
0,ARIMA,0.624492,0.705411,ARIMA,NaN,NaN,NaN,NaN,NaN,NaN
1,transformer,0.187262,0.532018,transformer_s192_p199_h4_k23_d0.1,192.0,199.0,4.0,0.1,64.0,2.0
2,informer,0.205611,0.511866,informer_s96_p199_h3_k23_d0.1,96.0,199.0,3.0,0.1,60.0,2.0
3,autoformer,0.180204,0.479890,autoformer_s96_p199_h5_k23_d0.2,96.0,199.0,5.0,0.2,80.0,2.0


In [ ]:
final_table.to_csv("results/final_model_comparison_table.csv", index=False)

In [24]:
from src.data.load_data import load_exchange_data
from src.data.lag_features import add_lags
from src.data.normalization import fit_scaler, transform_data
from src.experiments.run_experiment import run_experiment




In [ ]:
best_autoformer_config = {
    "model_type": "autoformer",
    "seq_len": 96,
    "pred_len": 199,
    "n_heads": 5,
    "dropout": 0.2,
    "d_model": 80,
    "enc_layers": 2,
    "kernel_size": 23,
    "batch_size": 16,
    "epochs": 20,
    "lr": 1e-4,
    "patience": 5,
    "residual": True,
    "univariate": False,
    "use_scaling": True,
    "save_path": "results/saved_models/autoformer_lag_test.pth"

}

In [26]:
baseline_config = best_autoformer_config.copy()
baseline_config["use_lags"] = False

baseline_result = run_experiment(baseline_config)

baseline_result

/Users/akarn/Documents/Exchange_curr_pred/src/models/autoformer.py:51: UserWarning: enable_nested_tensor is True, but self.use_nested_tensor is False because encoder_layer.self_attn.num_heads is odd
  self.encoder = nn.TransformerEncoder(


KeyError: 'save_path'

In [ ]:
lag_config = best_autoformer_config.copy()

lag_config["use_lags"] = True
lag_config["lags"] = [1,3,7]

lag_result = run_experiment(lag_config)

lag_result